File: twidere/src/main/java/org/mariotaku/twidere/activity/iface/IControlBarActivity.java
package org.mariotaku.twidere.activity.iface;

import android.animation.Animator;
import android.animation.Animator.AnimatorListener;
import android.animation.ObjectAnimator;
import android.util.Property;
import android.view.animation.DecelerateInterpolator;

/**
 * Created by mariotaku on 14/10/21.
 */
public interface IControlBarActivity {

    /**
     * @param offset 0: invisible, 1: visible
     */
    void setControlBarOffset(float offset);

    void setControlBarVisibleAnimate(boolean visible);

    void setControlBarVisibleAnimate(boolean visible, ControlBarShowHideHelper.ControlBarAnimationListener listener);

    float getControlBarOffset();

    int getControlBarHeight();

    void notifyControlBarOffsetChanged();

    void registerControlBarOffsetListener(ControlBarOffsetListener listener);

    void unregisterControlBarOffsetListener(ControlBarOffsetListener listener);

    interface ControlBarOffsetListener {
        void onControlBarOffsetChanged(IControlBarActivity activity, float offset);
    }

    final class ControlBarShowHideHelper {

        private static final long DURATION = 200L;

        private final IControlBarActivity activity;
        private int controlAnimationDirection;
        private ObjectAnimator currentControlAnimation;

        public ControlBarShowHideHelper(IControlBarActivity activity) {
            this.activity = activity;
        }

        private static class ControlBarOffsetProperty extends Property<IControlBarActivity, Float> {
            public static final ControlBarOffsetProperty SINGLETON = new ControlBarOffsetProperty();

            @Override
            public void set(IControlBarActivity object, Float value) {
                object.setControlBarOffset(value);
            }

            public ControlBarOffsetProperty() {
                super(Float.TYPE, null);
            }

            @Override
            public Float get(IControlBarActivity object) {
                return object.getControlBarOffset();
            }
        }

        public interface ControlBarAnimationListener {
            void onControlBarVisibleAnimationFinish(boolean visible);
        }

        public void setControlBarVisibleAnimate(boolean visible) {
            setControlBarVisibleAnimate(visible, null);
        }

        public void setControlBarVisibleAnimate(final boolean visible, final ControlBarAnimationListener listener) {
            final int newDirection = visible ? 1 : -1;
            if (controlAnimationDirection == newDirection) return;
            if (currentControlAnimation != null && controlAnimationDirection != 0) {
                currentControlAnimation.cancel();
                currentControlAnimation = null;
                controlAnimationDirection = newDirection;
            }
            final ObjectAnimator animator;
            final float offset = activity.getControlBarOffset();
            if (visible) {
                if (offset >= 1) return;
                animator = ObjectAnimator.ofFloat(activity, ControlBarOffsetProperty.SINGLETON, offset, 1);
            } else {
                if (offset <= 0) return;
                animator = ObjectAnimator.ofFloat(activity, ControlBarOffsetProperty.SINGLETON, offset, 0);
            }
            animator.setInterpolator(new DecelerateInterpolator());
            animator.addListener(new AnimatorListener() {
                @Override
                public void onAnimationStart(Animator animation) {
                }

                @Override
                public void onAnimationEnd(Animator animation) {
                    controlAnimationDirection = 0;
                    currentControlAnimation = null;
                    if (listener != null) {
                        listener.onControlBarVisibleAnimationFinish(visible);
                    }
                }

                @Override
                public void onAnimationCancel(Animator animation) {
                    controlAnimationDirection = 0;
                    currentControlAnimation = null;
                }

                @Override
                public void onAnimationRepeat(Animator animation) {

                }
            });
            animator.setDuration(DURATION);
            animator.start();
            currentControlAnimation = animator;
            controlAnimationDirection = newDirection;
        }
    }
}

File: twidere/src/main/kotlin/org/mariotaku/twidere/activity/iface/IControlBarActivity.kt
package org.mariotaku.twidere.activity.iface

import android.animation.Animator
import android.animation.Animator.AnimatorListener
import android.animation.ObjectAnimator
import android.util.Property
import android.view.animation.DecelerateInterpolator

/**
 * Created by mariotaku on 14/10/21.
 */
interface IControlBarActivity {

    /**
     * 0: invisible, 1: visible
     */
    var controlBarOffset: Float
        get() = 0f
        set(value) {}

    val controlBarHeight: Int get() = 0

    fun setControlBarVisibleAnimate(visible: Boolean, listener: ControlBarShowHideHelper.ControlBarAnimationListener? = null) {}

    fun registerControlBarOffsetListener(listener: ControlBarOffsetListener) {}

    fun unregisterControlBarOffsetListener(listener: ControlBarOffsetListener) {}

    fun notifyControlBarOffsetChanged() {}

    interface ControlBarOffsetListener {
        fun onControlBarOffsetChanged(activity: IControlBarActivity, offset: Float)
    }

    class ControlBarShowHideHelper(private val activity: IControlBarActivity) {
        private var controlAnimationDirection: Int = 0
        private var currentControlAnimation: ObjectAnimator? = null

        private object ControlBarOffsetProperty : Property<IControlBarActivity, Float>(Float::class.java, null) {

            override fun set(obj: IControlBarActivity, value: Float) {
                obj.controlBarOffset = (value)
            }

            override fun get(obj: IControlBarActivity): Float {
                return obj.controlBarOffset
            }

        }

        interface ControlBarAnimationListener {
            fun onControlBarVisibleAnimationStart(visible: Boolean) {}
            fun onControlBarVisibleAnimationFinish(visible: Boolean) {}
        }

        fun setControlBarVisibleAnimate(visible: Boolean, listener: ControlBarAnimationListener? = null) {
            val newDirection = if (visible) 1 else -1
            if (controlAnimationDirection == newDirection) return
            if (currentControlAnimation != null && controlAnimationDirection != 0) {
                currentControlAnimation!!.cancel()
                currentControlAnimation = null
                controlAnimationDirection = newDirection
            }
            val animator: ObjectAnimator
            val offset = activity.controlBarOffset
            animator = if (visible) {
                if (offset >= 1) return
                ObjectAnimator.ofFloat(activity, ControlBarOffsetProperty, offset, 1f)
            } else {
                if (offset <= 0) return
                ObjectAnimator.ofFloat(activity, ControlBarOffsetProperty, offset, 0f)
            }
            animator.interpolator = DecelerateInterpolator()
            animator.addListener(object : AnimatorListener {
                override fun onAnimationStart(animation: Animator) {
                    listener?.onControlBarVisibleAnimationStart(visible)
                }

                override fun onAnimationEnd(animation: Animator) {
                    controlAnimationDirection = 0
                    currentControlAnimation = null
                    listener?.onControlBarVisibleAnimationFinish(visible)
                }

                override fun onAnimationCancel(animation: Animator) {
                    controlAnimationDirection = 0
                    currentControlAnimation = null
                }

                override fun onAnimationRepeat(animation: Animator) {

                }
            })
            animator.duration = DURATION
            animator.start()
            currentControlAnimation = animator
            controlAnimationDirection = newDirection
        }

        companion object {

            private const val DURATION = 200L
        }
    }
}
